jetty-core/jetty-compression/jetty-compression-server/src/main/java/org/eclipse/jetty/compression/server/DynamicCompressionHandler.java

```diff
@@ -0,0 +1,252 @@
+                    {
+                        for (String value: values)
+                        {
+                            String lvalue = StringUtil.asciiToLowerCase(value);
+                            // only track encodings that are supported by this handler
+                            if (supportedEncodings.containsKey(lvalue))
+                            {
+                                if (requestAcceptEncoding == null)
+                                    requestAcceptEncoding = new ArrayList<>();
+                                requestAcceptEncoding.add(lvalue);
+                            }
+                        }
+                    }
+                }
+                case IF_MATCH, IF_NONE_MATCH ->
+                {
+                    etagMatches |= field.getValue().contains(EtagUtils.ETAG_SEPARATOR);
+                }
+            }
+        }
+
+        String decompressEncoding = config.getDecompressionEncoding(requestContentEncoding, request, pathInContext);
+        String compressEncoding = config.getCompressionEncoding(requestAcceptEncoding, request, pathInContext);
+
+        if (LOG.isDebugEnabled())
+        {
+            LOG.debug("Request[{}] Content-Encoding={}, Accept-Encoding={}, decompressEncoding={}, compressEncoding={}",
+                request, requestContentEncoding, requestAcceptEncoding, decompressEncoding, compressEncoding);
+        }
+
+        // Can we skip looking at the request and wrapping request or response?
+        if (decompressEncoding == null && compressEncoding == null)
+        {
+            if (LOG.isDebugEnabled())
+                LOG.debug("Skipping Compression and Decompression: no request encoding matches");
+            // No need for a Vary header, as we will never deflate
+            return next.handle(request, response, callback);
+        }
+
+        Request decompressionRequest = request;
+        Response compressionResponse = response;
+        Callback compressionCallback = callback;
+
+        // We need to wrap the request IFF we are inflating or have seen etags with compression separators
+        if (decompressEncoding != null || etagMatches)
+        {
+            decompressionRequest = newDecompressionRequest(request, decompressEncoding, config);
+        }
+
+        if (compressEncoding != null && config.getVary() != null)
+        {
+            // The response may vary based on the presence or lack of Accept-Encoding.
+            response.getHeaders().ensureField(config.getVary());
+        }
+
+        // Wrap the response and callback IFF we can be deflated and will try to deflate
+        if (compressEncoding != null)
+        {
+            Response compression = newCompressionResponse(this, request, response, callback, compressEncoding, config);
+            compressionResponse = compression;
+            if (compression instanceof Callback dynamicCallback)
+                compressionCallback = dynamicCallback;
+        }
```

Comment:
niggle:

```suggestion
        // Wrap the response and callback IFF we can be deflated and will try to deflate
        if (compressEncoding != null
        {
            // The response may vary based on the presence or lack of Accept-Encoding.
            if (config.getVary() != null)
                response.getHeaders().ensureField(config.getVary());
        
            Response compression = newCompressionResponse(this, request, response, callback, compressEncoding, config);
            compressionResponse = compression;
            if (compression instanceof Callback dynamicCallback)
                compressionCallback = dynamicCallback;
        }
```

Reply:
My reading of `Vary` header is that it should be present even if the content might be compressed at some other time, not just if it is actually compressing content.